### ISSUE TITLE
GEOS Lib doesn't support desk accessories

### ISSUE THREAD

**Issue by xlar54** [cc65/cc65#823]
Im already working on adding this functionality, so this is a ticket placeholder.

**Comment by xlar54** [comment]
Oliver - is there any way to get the length of the final compiled code?  Desk accessories in GEOS require knowledge of the exact file size so that the proper data can be restored.  I have it working if I hardcode the file start at $0400 and end as $5fff, which is the entire app space, but it would be more efficient if I knew the exact file size after compilation and linking.

**Comment by oliverschmidt** [comment]
I haven't seen a scenario so far where this wasn't possible. However "final compiled code" is too vague. It would be best if you could post the linker config file you use and point out which part(s) of the binary you consider the "final compiled code" by means of memory areas (or segments).

**Comment by xlar54** [comment]
Im using the default geos-cbm.cfg (in cfg folder) without any changes.   But let me study up more on that and see if I can figure it out

**Comment by oliverschmidt** [comment]
Okay, then just an example to help you looking at the right direction:
https://github.com/cc65/cc65/blob/master/cfg/apple2.cfg#L17 contains `define = yes` which allows https://github.com/cc65/cc65/blob/master/libsrc/apple2/exehdr.s#L10 to import `__MAIN_START__` and `__MAIN_LAST__`. https://github.com/cc65/cc65/blob/master/libsrc/apple2/exehdr.s#L15 then computes the size.

**Comment by xlar54** [comment]
Got it, thank you!

**Comment by xlar54** [comment]
Ok, the problem is that grc65 is the resource compiler used to generate the header required for GEOS.  It has no knowledge of the ca65 output of the main program, but its what needs to know the binary file length.  Will have to mull this one over a bit.  The chicken/egg situation is that the resource compiler runs *before* the main application compilation.  Must be one of the reasons the original GEOSLib author didnt include this functionality.

**Comment by xlar54** [comment]
Yeah, youre right... I see now.  So now, Im just trying to figure out which symbols I need to use.  Thanks again Oliver.  Getting closer.

One thing about it - with cc65, dont bother trying to resist the config files.  You're going to have to learn them and the linker at some point.  Ive been dancing around them and now Im learning them.

**Comment by xlar54** [comment]
ok, gonna punt on this one for now.  Tried various symbols and what made sense to me.  If anyone wants to try furthering this along,  attached is the main.c of the grc compiler that needs the fix.  The hardcoded $5fff needs to be the full binary size of the desk accessory, instead of the full app space.  The hardcoded value works, but naturally swapping in and out the entire space is slow and unnecessay.  The rest of my changes are in my own branch:  https://github.com/xlar54/cc65/tree/geos-deskacc-support

![capture](https://user-images.githubusercontent.com/1909642/50540260-1c367480-0b54-11e9-906f-cfa0bb0e1db6.PNG)

The below zip attachment contains a very simple desk accessory that does nothing but open up an overlapping window (without a frame) and prints some text.  Hitting enter should close it and return back to the application.  (Either the desktop or geopaint for instance)

[test.zip](https://github.com/cc65/cc65/files/2716542/test.zip)

Thanks to Oliver and Groepaz for their help, but alas properly calculating the binary size and putting it in this resource file just isnt working out for me.

Oliver, since there is a way to do it, and I doubt theres a great deal of interest in finishing it, if you suggest we close this issue, it will at least contain the info if someone ever wants to pick it back up.

**Comment by greg-king5** [comment]
When DAs are built, `__OVERLAYADDR__` must be redefined (on the command line).  And, that's the import name that must be used in the resource info block:
```asm
.addr __VLIR0_START__, __OVERLAYADDR__, __STARTUP_RUN__
```
It must be one location higher than the highest address that you want to save when loading the DA.  You must provide enough room for the stack, the heap (if you use it), and the BSS segment.  You always will need to refer to a generated map file.

Two command lines can generate everything.  The first one gets the map file.  The second one sets the address:
```sh
cl65 -t geos-cbm -O -m test.map -o test.cvt testres.grc test.c
cl65 -t geos-cbm -O -Wl -D,__OVERLAYADDR__=$2600 -o test.cvt testres.o test.o
```

A note about "test.c":
Don't PEEK()/POKE() GEOS's pseudo-registers.  They're defined as variables, not addresses.
```c
recoveryFlag = r10L;
```

**Comment by xlar54** [comment]
Im fairly confident that DAs are never VLIR files.  According to the documentation, the primary difference between apps and DAs is that apps can be VLIR with overlays to be much larger applications, so the end address in the header isnt required.  But DAs must have the ending address so that range of memory can be swapped in and out without disrupting the calling application.  On the 64, it creates a file called SWAPFILE on disk and pushes from start to end to disk, then loads the DA.  (On the 128, it pushes the memory space to block 1).  Not sure about GEOS Apple, but I would imagine it too writes to a disk swap file.  If DAs were VLIRs, then you would have to track much more than just start and end locations.  Attached is some sample DA code from geoProgrammer with some helpful comments.  Unfortunately, I cant seem to locate many other code examples.
[Desk Accessory Sample Code.zip](https://github.com/cc65/cc65/files/2717476/Desk.Accessory.Sample.Code.zip)

**Comment by oliverschmidt** [comment]
Just to avoid a communication deadlock: Even if there's no risk of a DA being a VLIR file I'd like to have a new linker symbol rather than `__OVERLAYADDR__` "reused".

> Oliver, since there is a way to do it, and I doubt theres a great deal of interest in finishing it, if you suggest we close this issue, it will at least contain the info if someone ever wants to pick it back up. Hate abandoning it after banging my head against the wall for a few days, and am trying to learn and contribute on what appears to be small things to help out.

I fully agree with you. I'd like to see support for DAs being finalized - instead of this pr closed. What info do you need to get it done?

**Comment by xlar54** [comment]
Let me look at it again over the weekend, and see if the ideas greg brought forward will help (with new linker symbol).  Ill keep you guys posted.

**Comment by greg-king5** [comment]
Commit 2959ade6e6c684d9982047bcc2e51717bc4dfee8 added `__HIMEM__` to the GEOS configure files.  Use that name instead of `__OVERLAYADDR__` when building DAs (it's shorter and easier to remember).

**Comment by polluks** [comment]
How about a DA section in geos.sgml?

**Comment by xlar54** [comment]
been awhile since I last checked in on this.  But looking back, I regret my earlier thought and agree with Oliver.  Its definitely solvable, so would be nice to seal it up.  If i can clear some time soon, Ill take another stab at it.
